### PR TITLE
2023 03 17 switch to clap

### PR DESCRIPTION
**Pull request by pdaoust**
### Summary

structopt 0.3 is woefully out of date. The biggest reason I switched to clap (which eventually integrated and subsumed structopt) was that multi-line arg and subcommand comments were being truncated to the first line.

Not knowing enough about how things might have changed since the version of clap that came with structopt, it may have introduced some subtle changes.

There is one **breaking change** -- the short arg for `hc sandbox --holochain-path` has changed from `-h` to `-H` because it clashed with the short arg for `--help`. I guess this was always a conflict; it just never got reported as an error with structopt 0.3. So I guess it should be considered a bugfix too.

**Not included**: I've only updated `hc` and its subcommands to clap; `holochain` and other binaries remain unchanged.

### TODO:
- [x] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

**Comment by pdaoust** [comment]
Note: since I'm new to clap / structopt, I don't quite understand all the bullet points in the "subtle changes (i.e., compiler won't catch)" section of [clap 4's changelog](https://github.com/clap-rs/clap/blob/master/CHANGELOG.md#breaking-changes-1). Would someone with more knowledge than I review that list?

File: crates/hc/src/lib.rs
Comment (ThetaSinner) [inline crates/hc/src/lib.rs:103]: Could the `impl` be moved down a level to the inner enum rather than the struct so this stays as `cmd.run().await`? To be consistent with the sandbox command on the next line

**Review by maackle — APPROVED** [review]
Thanks!

**Comment by jost-s** [comment]
Same for `hc s list`

**Comment by jost-s** [comment]
As a follow-up to this upgrade, we can consider adding shell autocompletion with https://github.com/clap-rs/clap/tree/master/clap_complete

File: crates/hc/src/lib.rs
Comment (pdaoust) [reply crates/hc/src/lib.rs:103]: sure, that'd simplify things. I'll do that today and push changes.

I'm learning so much about Rust!

**Comment by pdaoust** [comment]
> I'm getting an error when calling `hc s call`

crap, I thought I fixed that. I'm also wondering about the UX of shortened subcommands -- from the CLI's perspective `s` is unambiguously `sandbox` because `scaffold` is an external subcommand, but from the dev-user's perspective there are two possible things that `s` could mean and it's not clear why it would default to `sandbox`.

Not sure what a solution to that could be.

**Comment by pdaoust** [comment]
Fixed the issue with `hc s call`; moving on to refactoring the other stuff.

**Comment by pdaoust** [comment]
@jost-s @maackle I refactored the impl stuff; could you check the latest commit 00fbfaa ? don't know if it's well-structured/idiomatic.

File: crates/hc/src/lib.rs
Comment (maackle) [reply crates/hc/src/lib.rs:103]: I was agreeing with it being nice for this to just be `cmd.run()`, which I guess would actually require an impl on `cmd` in each of these cases. Maybe the better way to make it consistent is to remove that impl on HcSandbox so it's `cmd.command.run()` there too! Either way it's very minor, I'm ok with it as is.

**Comment by jost-s** [comment]
> @jost-s @maackle I refactored the impl stuff; could you check the latest commit [00fbfaa](https://github.com/holochain/holochain/commit/00fbfaaf7f607cf52841e4e52f4dd09d084c0953) ? don't know if it's well-structured/idiomatic.

All good from my end

File: crates/hc/src/lib.rs
Comment (pdaoust) [reply crates/hc/src/lib.rs]: I opted to put `run` on the weird wrapper structs I had to create for compatibility with the newest clap. The only exception is the three bundling subcommands, because recursive packing calls `HcDnaBundleSubcommand::Pack.run()`, and it seemed silly to create a `HcDnaBundle` instance just to stuff the subcommand in there and call `bundle.run()` on it. So `HcDnaBundle::run()` exists for consistency's sake, but it just calls `self::subcommand.run()`.

Incidentally, I also renamed `*Command` to `*Subcommand`, cuz it seemed more intuitive.

**Comment by pdaoust** [comment]
@jost-s merging is blocked on the bug you found; could you resolve the change request? 🙏🏼

**Comment by jost-s** [comment]
@pdaoust 

> Same for `hc s list`

This still doesn't work, I get an error:

```sh
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `u8`,
 right: `usize`: Argument `verbose`'s selected action Count contradicts `value_parser` (ValueParser::other(usize))', /Users/jost/Desktop/holochain/holochain/.cargo/registry/src/github.com-1ecc6299db9ec823/clap_builder-4.1.14/src/builder/debug_asserts.rs:696:9
 ```

Another command that's somehow lost is `hc --version`.

```sh
nix: holochain 2023-03-17-switch-to-clap $ target/debug/hc --version
error: unexpected argument '--version' found

Usage: hc <COMMAND>

For more information, try '--help'.
```

Please have another look.

**Comment by pdaoust** [comment]
@jost-s I just fixed both of those things, but it makes me very concerned that we're missing other edge cases with zero test coverage on the `hc` CLI.

**Comment by jost-s** [comment]
> @jost-s I just fixed both of those things, but it makes me very concerned that we're missing other new bugs post-clap-upgrade, with zero test coverage on the `hc` CLI.

I understand, yes. I don't know why there aren't test for the commands.

I've tested the two commands anyway and they work now.

**Comment by pdaoust** [comment]
@jost-s okay then, shall we merge this and hope for the best? If so, could you mark your request for changes as resolved?

**Review by jost-s — APPROVED** [review]

**Comment by jost-s** [comment]
You bet.

Quoting Michael here:
> If there are any subtle bugs, it seems worth it to have this upgrade anyway. From a glance, it seems like it'll be fine though.

As the Irish would say "sure it'll be grand!"

**Comment by pdaoust** [comment]
@maackle just getting back into the office -- do you want to block merging on this smoke test thing? I have absolutely no clue how to write those :D (but I have discovered that I can learn these things...)

**Comment by ThetaSinner** [comment]
@pdaoust For the `sandbox` sub-command I'm looking at getting the existing tests (out-of-date, haven't been run for a long time) into a good state to be run on CI here https://github.com/holochain/holochain/pull/2196. It's giving me some trouble though since launching `holochain` isn't as straightforward as I'd hoped. For the other commands in `hc` though, these tests might have some useful pieces that you can grab

**Comment by maackle** [comment]
@pdaoust I won't block. I'm inclined to say let's get smoke tests going before merging this, it seems that the risk of oft-used commands crashing is greater than the benefit of fixing the one subcommand you did, though I'm not close enough to the problem to really make that judgement. Maybe there can be a collaboration on those smoke tests, doesn't have to be only you, though I don't know who else has the bandwidth.

**Comment by jost-s** [comment]
I'd be happy to dedicate some time to these tests.

**Comment by pdaoust** [comment]
@maackle happy to hold onto this until there are smoke tests; thanks @jost-s for offering to do that. @maackle the urgency of this PR is even lower than fixing a subcommand -- it's just about improving helptext output (and in fact it doesn't change anything in that regard -- I mistakenly thought that truncation of multi-line subcommand helptext was a bug in structopt 0.3, whereas it's actually a feature, I guess.)

The only two pressing reasons to merge this are:

* it accidentally fixes #1423 (really should have separated it out into its own PR against `develop`)
* #2100 depends on it, and I'd really love to get #2100 into 0.2

**Comment by ThetaSinner** [comment]
@pdaoust The tests I was working on for `hc-sandbox` are on develop now from #2196. We might want some more coverage in those but do we think those are enough to proceed with this PR? There's more structopt usage going into develop and I'd be keen to get this in before we have too much more

**Comment by pdaoust** [comment]
@ThetaSinner I'm going to defer to @jost-s and @maackle re: whether this is sufficient -- I am absolutely clueless 😅  but I also am itching to close this.

**Comment by maackle** [comment]
Nice, I didn't see the tests @ThetaSinner wrote. Now that we have some tests and a framework for writing more of them it feels plausible to merge this.

Unfortunately this will affect https://github.com/holochain/holochain/pull/2265 but better sooner than later.

**Comment by pdaoust** [comment]
@maackle it looks like it was a fairly straightforward merge with `develop`, which contained #2265 (he said, before CI was finished).

**Pull request merged**
